core/src/main/scala/org/apache/spark/util/BestEffortLazyVal.scala

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.util
+
+/**
+ * Construct to lazily initialize a variable, re-computation is allowed if multiple threads are
+ * trying to initialize it concurrently.
+ * This may be helpful for avoiding deadlocks in certain scenarios.
+ *
+ * @note
+ * This helper class has additional requirements on the compute function:
+ *   1) The compute function MUST not return null;
+ *   2) The compute function MUST be deterministic.
+ *
+ * @note
+ *   Scala 3 uses a different implementation of lazy vals which doesn't have this problem.
```

Comment:
do we still need this note? scala 3 lazy val still use locks.

Reply:
I think scala 3 lazy won't use lock

https://docs.scala-lang.org/scala3/reference/changed-features/lazy-vals-init.html

Reply:
it is also based on CAS

Reply:
does it guarantee exactly once? If yes we should probably copy the implementation into Spark.

Reply:
I'm not sure whether it could be clearly moved to a util class

https://github.com/scala/scala3/blob/main/compiler/src/dotty/tools/dotc/transform/LazyVals.scala
https://github.com/scala/scala3/commit/b9d556574e29820b31f7183ea3e5f17fbe0dc627
https://github.com/scala/scala3/commit/3ef3a199d8a9d422a51150a23f5ed31fbc01024e

Reply:
OK so it's kind of a complicated implementation which lock-free and exactly once. Maybe we should have this simple version first as exactly once is not a hard requirement here.